### PR TITLE
WIP: Fix tests in case of running each relative test file.

### PR DESCRIPTION
**Pull request by junaruga**
## What was the end-user or developer problem that led to this PR?

Some tests fails when running with a specific relative test file. We should avoid using `__FILE__` in each `test_*` method, because a value of `__FILE__` in the `test_*.rb` is different executed from between `rake test` and `ruby -Ilib:test:bundler/lib test/path/to/test_*.rb`.

When running `rake test`, the value of `__FILE__` is the absolute path. When running `ruby -Ilib:test:bundler/lib test/rubygems/test_foo.rb`, the value of `__FILE__` is the relative path. Note that in the `test_*` method, the working directory (`Dir.pwd`) is `/path/to/tmp/test_rubygems_*`.

```
$ ruby -Ilib:test:bundler/lib test/rubygems/test_gem_ext_cargo_builder.rb -n /test_build_staticlib/
...
Error: test_build_staticlib(TestGemExtCargoBuilder): Errno::ENOENT: No such file or directory @ rb_check_realpath_internal - /home/jaruga/var/git/rubygems/tmp/test_rubygems_20220328-1989428-x5pim5/test/rubygems/test_gem_ext_cargo_builder/rust_ruby_example
...
test/rubygems/test_gem_ext_cargo_builder.rb:27:in `setup_rust_gem'
...
```

You can see https://github.com/rubygems/rubygems/issues/5411 for details.

## What is your fix for the problem, implemented in this PR?

I did set static file paths by not using `__FILE__`. Add a new test to run each relative test file isolatedly.

I am still working on it to fix other test files to pass the case of running each relative test file.

## Make sure the following tasks are checked

- [X] Describe the problem / feature
- [ ] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [ ] Write code to solve the problem
- [ ] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

### CODE REVIEW

**Comment by junaruga** [comment]
Any maintainers, could you review about the way to fix this issue for the current PR? If you agree on this way, I will work on other test files that have this issue too. The files maybe be the ones listed at https://github.com/rubygems/rubygems/issues/5411#issuecomment-1081921621 .

File: .github/workflows/ubuntu-rubygems.yml [hunk 1]
```diff
@@ -44,6 +44,9 @@ jobs:
       - name: Run Test
         run: rake test
         if: "!startsWith(matrix.ruby.name, 'truffleruby') && !startsWith(matrix.ruby.name, 'jruby')"
+      - name: Run Test isolatedly
+        run: rake test:isolated
+        if: "!startsWith(matrix.ruby.name, 'truffleruby') && !startsWith(matrix.ruby.name, 'jruby')"
```
Comment (junaruga) [inline .github/workflows/ubuntu-rubygems.yml:49]: We may not need to run this test in all the CRubies. Running on the latest stable CRuby might be enough good like this.

```
diff --git a/.github/workflows/ubuntu-rubygems.yml b/.github/workflows/ubuntu-rubygems.yml
index 5bc7ff8365..29cf6440f7 100644
--- a/.github/workflows/ubuntu-rubygems.yml
+++ b/.github/workflows/ubuntu-rubygems.yml
@@ -46,7 +46,7 @@ jobs:
         if: "!startsWith(matrix.ruby.name, 'truffleruby') && !startsWith(matrix.ruby.name, 'jruby')"
       - name: Run Test isolatedly
         run: rake test:isolated
-        if: "!startsWith(matrix.ruby.name, 'truffleruby') && !startsWith(matrix.ruby.name, 'jruby')"
+        if: "!startsWith(matrix.ruby.name, '3.1')"
       - name: Run Test (JRuby)
         run: JRUBY_OPTS=--debug rake test
         if: startsWith(matrix.ruby.name, 'jruby')
```

**Comment by simi** [comment]
> Any maintainers, could you review about the way to fix this issue for the current PR? If you agree on this way, I will work on other test files that have this issue too. The files maybe be the ones listed at [#5411 (comment)](https://github.com/rubygems/rubygems/issues/5411#issuecomment-1081921621) .

Hello. I'm ok with this solution. This blocked me to continue working on Rust builder improvements.

I'm not super sure we need that isolated setup on CI. @deivid-rodriguez is CI overlord and can make decision. Isn't there any other way to enforce `__FILE__` is not used in specs?

File: test/rubygems/path.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+require 'pathname'
+
+def source_root
```
Comment (simi) [inline test/rubygems/path.rb:5]: What about to keep it in module (similar to Bundler setup) to not "pollute" global (actually Kernel) "namespace"?

**Comment by deivid-rodriguez** [comment]
Let me have a look at this now.

**Review by deivid-rodriguez — COMMENTED** [review]
I think this is due to https://bugs.ruby-lang.org/issues/16978 (`__FILE__` not using realpaths), which is not an easy fix in a backwards compatible way as many scripts out there rely on things like

```ruby
if __FILE__ == $0
  do_something
end
```

The fix proposed here (rely on `__dir__` rather than on `__FILE__`) makes sense to me.

But I think we can update locations in place instead of adding a helper file and enforce proper `__dir__` usages by enabling the [`Style/ExpandPathArguments` RuboCop cop](https://docs.rubocop.org/rubocop/cops_style.html#styleexpandpatharguments).

Regarding a test that runs test files in isolation, it seems fine to me too, and I agree it's enough to run it only on the latest Ruby.

**Pull request closed without merging**